Errors when trying to install NEST on Saga

Hello.
When trying to install NEST 2.20.0 on the Norwegian HPC Saga I get:
"There were errors detected during the run of the NEST test suite!"

Here's my install script:

module purge
module load GSL/2.6-GCC-8.3.0
module load Python/3.7.4-GCCcore-8.3.0
module load SciPy-bundle/2019.10-foss-2019b-Python-3.7.4
module load CMake/3.12.1
module list
cmake -DCMAKE_INSTALL_PREFIX=$HOME/nest-2.20.0 \
-Dwith-python=3 \
 ..
make -j
make install
make installcheck

I attach the report directory.
[reports.zip](https://github.com/nest/nest-simulator/files/4220619/reports.zip)

@ricardomurphy Thanks for reporting this. The failing test is
```
FAIL: test_nn_pre_centered_synapse (nest.tests.test_stdp_nn_synapses.STDPNNSynapsesTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/cluster/home/ricardom/nest-2.20.0/lib64/python3.7/site-packages/nest/tests/test_stdp_nn_synapses.py", line 292, in test_nn_pre_centered_synapse
    self.do_nest_simulation_and_compare_to_reproduced_weight("nn_pre-centered")
  File "/cluster/home/ricardom/nest-2.20.0/lib64/python3.7/site-packages/nest/tests/test_stdp_nn_synapses.py", line 84, in do_nest_simulation_and_compare_to_reproduced_weight
    weight_by_nest, weight_reproduced_independently))
AssertionError: 0.32057074514327416 != 0.8197400647350731 within 7 places (0.49916931959179894 difference) : stdp_nn_pre-centered_synapse test: Resulting synaptic weight 8.197401e-01 differs from expected 3.205707e-01
```
I cannot reproduce this failure on my laptop at the moment, but we will investigate.

Could you run the test a few times over as follows and report if you always get the same error with the same values (repeating the last line):
```
cd <install_dir>
source bin/nest_vars.sh
nosetests lib/python3.7/site-packages/nest/tests/test_stdp_nn_synapses.py 
```

Hi.
I'm afraid I got errors when I tried to run the test:

[ricardom@login-1.SAGA ~/nest-2.20.0]$ source bin/nest_vars.sh
[ricardom@login-1.SAGA ~/nest-2.20.0]$ nosetests lib/python3.7/site-packages/nest/tests/test_stdp_nn_synapses.py
E
======================================================================
ERROR: Failure: ImportError (No module named test_stdp_nn_synapses)
----------------------------------------------------------------------
Traceback (most recent call last):
 File "/usr/lib/python2.7/site-packages/nose/loader.py", line 418, in loadTestsFromName
 addr.filename, addr.module)
 File "/usr/lib/python2.7/site-packages/nose/importer.py", line 47, in importFromPath
 return self.importFromDir(dir_path, fqname)
 File "/usr/lib/python2.7/site-packages/nose/importer.py", line 79, in importFromDir
 fh, filename, desc = find_module(part, path)
ImportError: No module named test_stdp_nn_synapses

----------------------------------------------------------------------
Ran 1 test in 0.000s

FAILED (errors=1)
[ricardom@login-1.SAGA ~/nest-2.20.0]$


___________________________________

Ricardo Murphy,
Sinsenveien 5A,
0572 Oslo,
Norway


On Tue, Feb 18, 2020, at 23:25, Hans Ekkehard Plesser wrote:
> @ricardomurphy <https://github.com/ricardomurphy> Thanks for reporting this. The failing test is

> `FAIL: test_nn_pre_centered_synapse (nest.tests.test_stdp_nn_synapses.STDPNNSynapsesTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/cluster/home/ricardom/nest-2.20.0/lib64/python3.7/site-packages/nest/tests/test_stdp_nn_synapses.py", line 292, in test_nn_pre_centered_synapse
    self.do_nest_simulation_and_compare_to_reproduced_weight("nn_pre-centered")
  File "/cluster/home/ricardom/nest-2.20.0/lib64/python3.7/site-packages/nest/tests/test_stdp_nn_synapses.py", line 84, in do_nest_simulation_and_compare_to_reproduced_weight
    weight_by_nest, weight_reproduced_independently))
AssertionError: 0.32057074514327416 != 0.8197400647350731 within 7 places (0.49916931959179894 difference) : stdp_nn_pre-centered_synapse test: Resulting synaptic weight 8.197401e-01 differs from expected 3.205707e-01
`
> I cannot reproduce this failure on my laptop at the moment, but we will investigate.

> Could you run the test a few times over as follows and report if you always get the same error with the same values (repeating the last line):

> `cd <install_dir>
source bin/nest_vars.sh
nosetests lib/python3.7/site-packages/nest/tests/test_stdp_nn_synapses.py 
`

> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/nest/nest-simulator/issues/1435?email_source=notifications&email_token=AFM5BCSKBFZLQOWLXEHJMI3RDRN4DA5CNFSM4KXIZXYKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEMFSYGY#issuecomment-587934747>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AFM5BCURAA25OAN5XYDLUY3RDRN4DANCNFSM4KXIZXYA>.

Thanks, that worked. Here are my loaded modules followed by three runs of the test:

[ricardom@login-1.SAGA ~]$ module list

Currently Loaded Modules:
 1) StdEnv (S)
 2) CMake/3.12.1
 3) GCCcore/8.3.0
 4) zlib/1.2.11-GCCcore-8.3.0 (H)
 5) binutils/2.32-GCCcore-8.3.0 (H)
 6) GCC/8.3.0
 7) GSL/2.6-GCC-8.3.0
 8) bzip2/1.0.8-GCCcore-8.3.0 (H)
 9) ncurses/6.1-GCCcore-8.3.0 (H)
10) libreadline/8.0-GCCcore-8.3.0 (H)
11) XZ/5.2.4-GCCcore-8.3.0 (H)
12) GMP/6.1.2-GCCcore-8.3.0 (H)
13) libffi/3.2.1-GCCcore-8.3.0 (H)
14) Python/3.7.4-GCCcore-8.3.0
15) iccifort/2019.5.281
16) impi/2018.5.288-iccifort-2019.5.281
17) iimpi/2019b
18) imkl/2019.5.281-iimpi-2019b
19) intel/2019b
20) SciPy-bundle/2019.10-intel-2019b-Python-3.7.4

 Where:
 S: Module is Sticky, requires --force to unload or purge
 H: Hidden Module

 cd nest-2.20.0
[ricardom@login-1.SAGA ~/nest-2.20.0]$ ls
bin  include  lib64  share
[ricardom@login-1.SAGA ~/nest-2.20.0]$ source bin/nest_vars.sh

[ricardom@login-1.SAGA ~/nest-2.20.0]$ nosetests lib64/python3.7/site-packages/nest/tests/test_stdp_nn_synapses.py
[INFO] [2020.2.19 11:41:0 /cluster/home/ricardom/nest-simulator-2.20.0/nestkernel/rng_manager.cpp:217 @ Network::create__rngs__] : Creating default RNGs
[INFO] [2020.2.19 11:41:0 /cluster/home/ricardom/nest-simulator-2.20.0/nestkernel/rng_manager.cpp:260 @ Network::create__grng__] : Creating new default global RNG

              -- N E S T --
  Copyright (C) 2004 The NEST Initiative

 Version: nest-2.20.0
 Built: Feb 18 2020 17:24:12

 This program is provided AS IS and comes with
 NO WARRANTY. See the file LICENSE for details.

 Problems or suggestions?
   Visit https://www.nest-simulator.org

 Type 'nest.help()' to find out more about NEST.

F..
======================================================================
FAIL: test_nn_pre_centered_synapse (nest.tests.test_stdp_nn_synapses.STDPNNSynapsesTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/cluster/home/ricardom/nest-2.20.0/lib64/python3.7/site-packages/nest/tests/test_stdp_nn_synapses.py", line 292, in test_nn_pre_centered_synapse
    self.do_nest_simulation_and_compare_to_reproduced_weight("nn_pre-centered")
  File "/cluster/home/ricardom/nest-2.20.0/lib64/python3.7/site-packages/nest/tests/test_stdp_nn_synapses.py", line 84, in do_nest_simulation_and_compare_to_reproduced_weight
    weight_by_nest, weight_reproduced_independently))
AssertionError: 0.32057074514327416 != 0.8197400647350731 within 7 places (0.49916931959179894 difference) : stdp_nn_pre-centered_synapse test: Resulting synaptic weight 8.197401e-01 differs from expected 3.205707e-01

----------------------------------------------------------------------
Ran 3 tests in 5.456s

FAILED (failures=1)
[ricardom@login-1.SAGA ~/nest-2.20.0]$


[ricardom@login-1.SAGA ~/nest-2.20.0]$ nosetests lib64/python3.7/site-packages/nest/tests/test_stdp_nn_synapses.py
[INFO] [2020.2.19 11:43:12 /cluster/home/ricardom/nest-simulator-2.20.0/nestkernel/rng_manager.cpp:217 @ Network::create__rngs__] : Creating default RNGs
[INFO] [2020.2.19 11:43:12 /cluster/home/ricardom/nest-simulator-2.20.0/nestkernel/rng_manager.cpp:260 @ Network::create__grng__] : Creating new default global RNG

              -- N E S T --
  Copyright (C) 2004 The NEST Initiative

 Version: nest-2.20.0
 Built: Feb 18 2020 17:24:12

 This program is provided AS IS and comes with
 NO WARRANTY. See the file LICENSE for details.

 Problems or suggestions?
   Visit https://www.nest-simulator.org

 Type 'nest.help()' to find out more about NEST.

F..
======================================================================
FAIL: test_nn_pre_centered_synapse (nest.tests.test_stdp_nn_synapses.STDPNNSynapsesTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/cluster/home/ricardom/nest-2.20.0/lib64/python3.7/site-packages/nest/tests/test_stdp_nn_synapses.py", line 292, in test_nn_pre_centered_synapse
    self.do_nest_simulation_and_compare_to_reproduced_weight("nn_pre-centered")
  File "/cluster/home/ricardom/nest-2.20.0/lib64/python3.7/site-packages/nest/tests/test_stdp_nn_synapses.py", line 84, in do_nest_simulation_and_compare_to_reproduced_weight
    weight_by_nest, weight_reproduced_independently))
AssertionError: 0.32057074514327416 != 0.8197400647350731 within 7 places (0.49916931959179894 difference) : stdp_nn_pre-centered_synapse test: Resulting synaptic weight 8.197401e-01 differs from expected 3.205707e-01

----------------------------------------------------------------------
Ran 3 tests in 5.031s

FAILED (failures=1)
[ricardom@login-1.SAGA ~/nest-2.20.0]$


nosetests lib64/python3.7/site-packages/nest/tests/test_stdp_nn_synapses.py
[INFO] [2020.2.19 11:45:57 /cluster/home/ricardom/nest-simulator-2.20.0/nestkernel/rng_manager.cpp:217 @ Network::create__rngs__] : Creating default RNGs
[INFO] [2020.2.19 11:45:57 /cluster/home/ricardom/nest-simulator-2.20.0/nestkernel/rng_manager.cpp:260 @ Network::create__grng__] : Creating new default global RNG

              -- N E S T --
  Copyright (C) 2004 The NEST Initiative

 Version: nest-2.20.0
 Built: Feb 18 2020 17:24:12

 This program is provided AS IS and comes with
 NO WARRANTY. See the file LICENSE for details.

 Problems or suggestions?
   Visit https://www.nest-simulator.org

 Type 'nest.help()' to find out more about NEST.

F..
======================================================================
FAIL: test_nn_pre_centered_synapse (nest.tests.test_stdp_nn_synapses.STDPNNSynapsesTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/cluster/home/ricardom/nest-2.20.0/lib64/python3.7/site-packages/nest/tests/test_stdp_nn_synapses.py", line 292, in test_nn_pre_centered_synapse
    self.do_nest_simulation_and_compare_to_reproduced_weight("nn_pre-centered")
  File "/cluster/home/ricardom/nest-2.20.0/lib64/python3.7/site-packages/nest/tests/test_stdp_nn_synapses.py", line 84, in do_nest_simulation_and_compare_to_reproduced_weight
    weight_by_nest, weight_reproduced_independently))
AssertionError: 0.32057074514327416 != 0.8197400647350731 within 7 places (0.49916931959179894 difference) : stdp_nn_pre-centered_synapse test: Resulting synaptic weight 8.197401e-01 differs from expected 3.205707e-01

----------------------------------------------------------------------
Ran 3 tests in 5.410s

FAILED (failures=1)
[ricardom@login-1.SAGA ~/nest-2.20.0]$


Ricardo

___________________________________

Dr. Ricardo Murphy,
University of Oslo,
Domus Medica,
Division of Physiology,
Pb. 1103, Blindern,
0317 Oslo


On Wed, Feb 19, 2020, at 10:46, Hans Ekkehard Plesser wrote:
> @ricardomurphy <https://github.com/ricardomurphy> The traceback shows `python2.7`, while you built NEST with 3.7. It looks like you may not have loaded all the modules you loaded in your build script before running the test now. Could you try again with all modules loaded?


> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/nest/nest-simulator/issues/1435?email_source=notifications&email_token=AFM5BCX6PXB4VFDHVY2ZC6TRDT5YRA5CNFSM4KXIZXYKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEMHCFSQ#issuecomment-588128970>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AFM5BCTH5OVH7SAARKQXXNDRDT5YRANCNFSM4KXIZXYA>.

I got the same error when compiled without MPI. However, when I compiled it with-mpi=ON, this error dissappeard, and just to make sure, executing

`
nosetests lib/python3.7/site-packages/nest/tests/test_stdp_nn_synapses.py
`

resulted in no error.
I wonder if @ricardomurphy  is trying to install without MPI support on a HPC cluster...

Below is out of the scope of this issue, however, when compiled with-mpi=ON gave the following errors:

> FAIL: testWithMPI (nest.tests.test_connect_all_patterns.TestConnectAllPatterns)
      ", ".join(failing_tests))
  AssertionError: False is not true : The following tests failed when executing with "mpirun -np 2 nosetests [script]": test_connect_all_to_all.py, test_connect_one_to_one.py, test_connect_fixed_indegree.py, test_connect_fixed_outdegree.py, test_connect_fixed_total_number.py, test_connect_pairwise_bernoulli.py
  FAIL: testWithMPI (nest.tests.test_sp.test_mpitests.TestStructuralPlasticityMPI)
      '[script]": {}'.format(failing_tests))
  AssertionError: False is not true : The following tests failed when executing with "mpirun -np 2 nosetests [script]": ['/opt/ohpc/pub/libs/gnu8/openmpi3/nest/2.20.0/lib64/python3.6/site-packages/nest/tests/test_sp/mpitest_issue_578_sp.py']

which resulted to no error when executed with

`
mpirun -np 2 nosetests [script]
`

Which I would assume that my installation is OK.

Hope it helps and also apply for @ricardomurphy case.

The module configuration was the same in both cases:

`module list`

> Currently Loaded Modules:
>   autotools
>   prun/1.3
>   gnu8/8.3.0
>   openmpi3/3.1.4
>   ohpc
>   cmake/3.15.4
>   gsl/2.6
>   fftw/3.3.8
>   openblas/0.3.7
>   python/3.6.6

--------------------------------------------------------------------------------
### CMake command and Configuration that led to the same error as @ricardomurphy 

```
cmake -DCMAKE_INSTALL_PREFIX:PATH=/home/nilton/opt/nest/2.20.0-1435 -Dwith-python=3 -DPYTHON_EXECUTABLE=`which python3.6` -DNOSETESTS=`which nosetests` -DCYTHON_EXECUTABLE=`which cython` ../src
```
NEST Configuration Summary
--------------------------------------------------------------------------------

Build type          : 
Target System       : Linux
Cross Compiling     : FALSE
C compiler          : GNU 8.3.0 (/opt/ohpc/pub/compiler/gcc/8.3.0/bin/gcc)
C compiler flags    :  -O2 -Wall -fopenmp -fdiagnostics-color=auto
C++ compiler        : GNU 8.3.0 (/opt/ohpc/pub/compiler/gcc/8.3.0/bin/c++)
C++ compiler flags  :  -std=c++11 -O2 -Wall -fopenmp -fdiagnostics-color=auto
Build dynamic       : ON

Built-in modules    : models;precise;topology
User modules        : None
Python bindings     : Yes (Python 3.6.8: /usr/bin/python3.6)
       Includes     : /usr/include/python3.6m
       Libraries    : /usr/lib64/libpython3.6m.so

Cython bindings     : Yes (Cython 0.29.6: /opt/ohpc/pub/libs/gnu8/python/3.6.6/bin/cython)
Use threading       : Yes (OpenMP: -fopenmp)
Use GSL             : Yes (GSL 2.6)
    Includes        : /opt/ohpc/pub/libs/gnu8/gsl/2.6/include
    Libraries       : /opt/ohpc/pub/libs/gnu8/gsl/2.6/lib/libgsl.so;/opt/ohpc/pub/libs/gnu8/gsl/2.6/lib/libgslcblas.so

Use Readline        : Yes (GNU Readline 6.2)
    Includes        : /usr/include
    Libraries       : /lib64/libreadline.so;/lib64/libncurses.so

Use libltdl         : Yes (LTDL 2.4.6)
    Includes        : /usr/include
    Libraries       : /lib64/libltdl.so

Use doxygen         : Yes (/usr/bin/doxygen)
                    : target `doc` available
Use MPI             : No
Use MUSIC           : No
Use libneurosim     : No
Use Boost           : No


```
make installcheck
```

> FAIL: test_nn_pre_centered_synapse (nest.tests.test_stdp_nn_synapses.STDPNNSynapsesTest)
>   AssertionError: 0.32057074514327416 != 0.8197400647350731 within 7 places : stdp_nn_pre-centered_synapse test: Resulting synaptic weight 8.197401e-01 differs from expected 3.205707e-01
>   FAILED (SKIP=9, failures=1)
> 

--------------------------------------------------------------------------------
### CMake command and Configuration with MPI
```
cmake -DCMAKE_INSTALL_PREFIX:PATH=/home/nilton/opt/nest/2.20.0-1435 -Dwith-python=3 -DPYTHON_EXECUTABLE=`which python3.6` -DNOSETESTS=`which nosetests` -DCYTHON_EXECUTABLE=`which cython` -Dwith-mpi=ON ../src
```
NEST Configuration Summary
--------------------------------------------------------------------------------

Build type          : 
Target System       : Linux
Cross Compiling     : FALSE
C compiler          : GNU 8.3.0 (/opt/ohpc/pub/compiler/gcc/8.3.0/bin/gcc)
C compiler flags    :  -O2 -Wall -fopenmp   -pthread -fdiagnostics-color=auto
C++ compiler        : GNU 8.3.0 (/opt/ohpc/pub/compiler/gcc/8.3.0/bin/c++)
C++ compiler flags  :  -std=c++11 -O2 -Wall -fopenmp -pthread -fdiagnostics-color=auto
Build dynamic       : ON

Built-in modules    : models;precise;topology
User modules        : None
Python bindings     : Yes (Python 3.6.8: /usr/bin/python3.6)
       Includes     : /usr/include/python3.6m
       Libraries    : /usr/lib64/libpython3.6m.so

Cython bindings     : Yes (Cython 0.29.6: /opt/ohpc/pub/libs/gnu8/python/3.6.6/bin/cython)
Use threading       : Yes (OpenMP: -fopenmp)
Use GSL             : Yes (GSL 2.6)
    Includes        : /opt/ohpc/pub/libs/gnu8/gsl/2.6/include
    Libraries       : /opt/ohpc/pub/libs/gnu8/gsl/2.6/lib/libgsl.so;/opt/ohpc/pub/libs/gnu8/gsl/2.6/lib/libgslcblas.so

Use Readline        : Yes (GNU Readline 6.2)
    Includes        : /usr/include
    Libraries       : /lib64/libreadline.so;/lib64/libncurses.so

Use libltdl         : Yes (LTDL 2.4.6)
    Includes        : /usr/include
    Libraries       : /lib64/libltdl.so

Use doxygen         : Yes (/usr/bin/doxygen)
                    : target `doc` available
Use MPI             : Yes (MPI: /opt/ohpc/pub/mpi/openmpi3-gnu8/3.1.4/bin/mpicxx)
    FLAGS           : -pthread
    Includes        : /opt/ohpc/pub/mpi/openmpi3-gnu8/3.1.4/include
    Link Flags      : -Wl,-rpath -Wl,/opt/ohpc/pub/mpi/openmpi3-gnu8/3.1.4/lib -Wl,--enable-new-dtags -pthread
    Libraries       : /opt/ohpc/pub/mpi/openmpi3-gnu8/3.1.4/lib/libmpi_cxx.so;/opt/ohpc/pub/mpi/openmpi3-gnu8/3.1.4/lib/libmpi.so

Use MUSIC           : No
Use libneurosim     : No
Use Boost           : No


```
make installcheck
```

> FAIL: testWithMPI (nest.tests.test_connect_all_patterns.TestConnectAllPatterns)
>       ", ".join(failing_tests))
>   AssertionError: False is not true : The following tests failed when executing with "mpirun -np 2 nosetests [script]": test_connect_all_to_all.py, test_connect_one_to_one.py, test_connect_fixed_indegree.py, test_connect_fixed_outdegree.py, test_connect_fixed_total_number.py, test_connect_pairwise_bernoulli.py
>   FAIL: testWithMPI (nest.tests.test_sp.test_mpitests.TestStructuralPlasticityMPI)
>       '[script]": {}'.format(failing_tests))
>   AssertionError: False is not true : The following tests failed when executing with "mpirun -np 2 nosetests [script]": ['/home/nilton/opt/nest/2.20.0/lib64/python3.6/site-packages/nest/tests/test_sp/mpitest_issue_578_sp.py']
>   FAILED (SKIP=7, failures=2)
> 
--------------------------------------------------------------------------------

As I said before, the only difference was to enabling MPI

--- a/models/stdp_nn_pre-centered_connection.h
+++ b/models/stdp_nn_pre-centered_connection.h
@@ -322,6 +322,7 @@ STDPNNPreCenteredConnection< targetidentifierT >::STDPNNPreCenteredConnection(
   , mu_plus_( rhs.mu_plus_ )
   , mu_minus_( rhs.mu_minus_ )
   , Wmax_( rhs.Wmax_ )
+  , Kplus_( rhs.Kplus_ )
   , t_lastspike_( rhs.t_lastspike_ )
 {
 }

